File: Week_02/429/n_ary_tree_level_order_traversal2.py
from typing import List


class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class Solution:
    def levelOrder(self, root: 'Node') -> List[List[int]]:
        res = []
        level = 0
        stack = [root]
        while stack:
            num = len(stack)
            for i in range(num):
                if stack[i] is not None:
                    if len(res) == level:
                        res.append([stack[i].val])
                    else:
                        res[level].append(stack[i].val)
                    for c in stack[i].children:
                        stack.append(c)
            stack = stack[num:]
            level += 1
        return res

File: Week_02/429/n_ary_tree_level_order_traversal2.go
package main

func levelOrder2(root *Node) [][]int {
	res := make([][]int, 0)
	stack := []*Node{root}
	level := 0
	for len(stack) > 0 {
		num := len(stack)
		for i := 0; i < num; i++ {
			if stack[i] != nil {
				if len(res) == level {
					res= append(res, []int{stack[i].Val})
				} else {
					res[level] = append(res[level], stack[i].Val)
				}
				for j := 0; j < len(stack[i].Children); j++ {
					stack = append(stack, stack[i].Children[j])
				}
			}
		}
		stack = stack[num:]
		level++
	}
	return res
}
